text/3263-pre-release-sticky.md

```diff
@@ -0,0 +1,115 @@
+
+This is an open problem since some time, [cargo #2222](https://github.com/rust-lang/cargo/issues/2222) but have receive some input in Oct 2021, specially in this [forum post](https://internals.rust-lang.org/t/changing-cargo-semver-compatibility-for-pre-releases/14820) where this proposition have been suggested.
+
+Finally, contrary to release version where most user benefits from having cargo pick the latest compatible version as default, there is little sense to do the same for pre-release version cause there are very likely be breaking change between all pre-release version. This is a trap that most user don't expect to happen (why does my project don't compile anymore ?!?). Pre-release version precedence of `Cargo.toml` should default to something more user friendly. User expect bugs from a pre-release but no one want their build suddenly break when a new pre-release is out.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default Cargo follow Semver precedence rules for release of a crate but will make no assumption for pre-release.
+
+```toml
+[dependencies.foo]
+version = "1.0.0-alpha.0"
+```
+
+will now be considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "=1.0.0-alpha.0"
+```
+
+while release version like:
+
+```toml
+[dependencies.foo]
+version = "1.0.0"
+```
+
+are considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "^1.0.0"
+```
+
+The latter doesn't change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Do it transparently
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ * This could break workflow
+ * It's make pre-release semver rules have a different behavior than release for Cargo.
+ * It's a change that affect the whole Rust ecosystem.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+In practice, do it transparently should not break any heathy workflow, user that put a pre-release version clearly want to use this version and if user have been using a most recent pre-release version not on purpose it's very unlikely their build would break. We could check the impact of implementing this transparently with a [crater](https://github.com/rust-lang/crater) run.
+
+To introduce this change we could also:
+
+  1. Use the `resolver` value version, bump it to `3`:
+   
+     This may be overkill but it's a naturel way to make change in the way semver rule are trated by Cargo.
+     
+  2. Use a separate value `pre-release-updates = "sticky" # or "default"`:
+
+     This is less overkill, but have the drawback to include a new value in `Cargo.toml` just for that.
+
+Both these solutions would need a new Rust edition to be introduce by default.
```

Comment:
The solutions don't need a new Rust edition to be introduced by default, the idea of backwards compatibility makes that necessary. So this is more about adding a (potentially temporary) way to opt in to the new behavior before it potentially gets enabled by default in the next edition.

Reply:
So `pre-release-updates` is maybe better than `resolver`.